instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ClientSslTest.groovy

```diff
@@ -0,0 +1,225 @@
+            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
+            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
+            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port
+          }
+        }
+        span(2) {
+          name "CONNECT"
+          kind INTERNAL
+          childOf span(0)
+          attributes {
+            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
+            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
+            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port
+            "${SemanticAttributes.NET_PEER_IP.key}" { it == null || it == "127.0.0.1" }
+          }
+        }
+        span(3) {
+          name "SSL handshake"
+          kind INTERNAL
+          childOf span(0)
+          status ERROR
+          // netty swallows the exception, it doesn't make any sense to hard-code the message
+          errorEventWithAnyMessage(SSLHandshakeException)
+          attributes {
+            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
+            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
+            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port
+            "${SemanticAttributes.NET_PEER_IP.key}" { it == null || it == "127.0.0.1" }
+          }
+        }
+      }
+    }
+
+    cleanup:
+    channel?.close()?.sync()
+  }
+
+  def "should successfully establish SSL handshake"() {
+    given:
+    def bootstrap = createBootstrap(eventLoopGroup)
+
+    def uri = server.resolveHttpsAddress("/success")
+    def request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, uri.toString(), Unpooled.EMPTY_BUFFER)
+    request.headers().set(HttpHeaderNames.HOST, uri.host)
+
+    when:
+    Channel channel = null
+    runWithSpan("parent") {
+      channel = bootstrap.connect(uri.host, uri.port).sync().channel()
+      def result = new CompletableFuture<Integer>()
+      channel.pipeline().addLast(new ClientHandler(result))
+      channel.writeAndFlush(request).get(10, TimeUnit.SECONDS)
+      result.get(10, TimeUnit.SECONDS)
+    }
+
+    then:
+    assertTraces(1) {
+      trace(0, 6) {
+        span(0) {
+          name "parent"
+        }
+        span(1) {
+          name "RESOLVE"
+          kind INTERNAL
+          childOf span(0)
+          attributes {
+            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
+            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
+            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port
+          }
+        }
+        span(2) {
+          name "CONNECT"
+          kind INTERNAL
+          childOf span(0)
+          attributes {
+            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
+            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
+            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port
+            "${SemanticAttributes.NET_PEER_IP.key}" { it == null || it == "127.0.0.1" }
+          }
+        }
+        span(3) {
+          name "SSL handshake"
+          kind INTERNAL
```

Comment:
should `RESOLVE`, `CONNECT` and `SSL handshake` be `CLIENT` spans? (users would need to enable `outgoing-span-suppression-by-type=true`, or we could consider making suppression-by-type the default)

sort of related, what do you think of converting the "error-only" spans into span events on the current span? this way those would still get captured even when using the "no nested CLIENT spans" setting

I can create an issue to discuss both of these further since they have broader impact than just this PR

Reply:
> I can create an issue to discuss both of these further since they have broader impact than just this PR

Yes, please! There's a lot of things going on here that we could discuss before establishing a "common pattern" for tracing these things.

> should `RESOLVE`, `CONNECT` and `SSL handshake` be `CLIENT` spans?

I think it makes sense for them to be `CLIENT` spans. I was following the current implementation, which used `INTERNAL` everywhere though.

> users would need to enable `outgoing-span-suppression-by-type=true`, or we could consider making suppression-by-type the default)

In plain netty the connection phase and actually sending the HTTP request are completely separate; so all those low-level spans are not the children of the `CLIENT` span. In case of "normal"/higher-level HTTP clients I think it'd make perfect sense for them to be nested under the HTTP `CLIENT` span.

> sort of related, what do you think of converting the "error-only" spans into span events on the current span? this way those would still get captured even when using the "no nested CLIENT spans" setting

Hmm, do we want that in all cases? Or only when nested spans are disabled?
Using spans for these operations (as opposing to events) enables gathering timing and exception information; e.g. you know that SSL handshake took 10 seconds and threw an exception, and it failed before the HTTP request was sent.

Reply:
👍 created issue to continue discussion #4617